apps/voting/contracts/Voting.sol

```diff
@@ -122,79 +130,74 @@ contract Voting is IForwarder, AragonApp {
 
         address msgSender = msg.sender;
         require(_delegate != msgSender, ERROR_SELF_DELEGATE);
 
-        address prevDelegate = delegates[msgSender];
+        address prevDelegate = delegates[msgSender].delegate;
         require(_delegate != prevDelegate, ERROR_DELEGATE_SAME_AS_PREV);
 
+        uint256 votingPower = token.balanceOfAt(msgSender, getBlockNumber64() - 1);
+        require(votingPower > 0, ERROR_NO_VOTING_POWER);
+
         if (prevDelegate != address(0)) {
             _removeDelegatedAddressFor(prevDelegate, msgSender);
         }
 
         _addDelegatedAddressFor(_delegate, msgSender);
-        delegates[msgSender] = _delegate;
+        delegates[msgSender] = Delegate(_delegate, uint96(delegatedVoters[_delegate].addresses.length - 1));
 
         emit DelegateSet(msgSender, prevDelegate, _delegate);
     }
 
-    /**
-     * TODO: Calculate gas spending and add hint for more efficient look up in the array if needed
-     */
     function removeDelegate() public {
         address msgSender = msg.sender;
-        address prevDelegate = delegates[msgSender];
+        address prevDelegate = delegates[msgSender].delegate;
         require(prevDelegate != address(0), ERROR_DELEGATE_NOT_SET);
 
         _removeDelegatedAddressFor(prevDelegate, msgSender);
-        delegates[msgSender] = address(0);
+        delegates[msgSender] = Delegate(address(0), 0);
 
         emit DelegateSet(msgSender, prevDelegate, address(0));
     }
 
     function _addDelegatedAddressFor(address _delegate, address _voter) internal {
-        require(_delegate != address(0), ERROR_ZERO_ADDRESS_PASSED);
-        require(_voter != address(0), ERROR_ZERO_ADDRESS_PASSED);
-
-        uint256 length = delegatedVoters[_delegate].addresses.length;
-        require(length < MAX_VOTERS_PER_DELEGATE, ERROR_MAX_DELEGATED_VOTERS_REACHED);
-
         delegatedVoters[_delegate].addresses.push(_voter);
     }
 
     function _removeDelegatedAddressFor(address _delegate, address _voter) internal {
-        require(_delegate != address(0), ERROR_ZERO_ADDRESS_PASSED);
-        require(_voter != address(0), ERROR_ZERO_ADDRESS_PASSED);
+        uint96 voterIndex = delegates[_voter].voterIndex;
 
         uint256 length = delegatedVoters[_delegate].addresses.length;
-        for (uint256 i = 0; i < length; i++) {
-            if (delegatedVoters[_delegate].addresses[i] == _voter) {
-                delegatedVoters[_delegate].addresses[i] = delegatedVoters[_delegate].addresses[length - 1];
-                delete delegatedVoters[_delegate].addresses[length - 1];
-                delegatedVoters[_delegate].addresses.length--;
-                break;
-            }
+        delegatedVoters[_delegate].addresses[voterIndex] = delegatedVoters[_delegate].addresses[length - 1];
+        delegates[delegatedVoters[_delegate].addresses[length - 1]].voterIndex = voterIndex;
+        delete delegatedVoters[_delegate].addresses[length - 1];
```

Comment:
needs a separate unittest checking "last delegate changed / last delegate removed" for single- and multiple-delegate lists